Add vectorized approximation of exp(x)

j-brotz opened the pull request:
This draft PR introduces an algorithm for computing an approximation of the exponential function in a vectorized fashion. We initially used this approach to accelerate our implementation of a microstructure evolution model [https://doi.org/10.1016/j.addma.2024.104380](https://doi.org/10.1016/j.addma.2024.104380 "Persistent link using digital object identifier"), and believe it could be a useful addition for other models as well. 

Before diving into the algorithm and implementation details, I would like to briefly comment on the performance. Benchmarking was somewhat tricky, and the exact speedups may depend on several factors. On my machine (AMD Ryzen 9 7940HS with AVX512 support), I measured an overall speedup of approximately:

- 18 using `VectorizedArray<double, 8>`
- 25 using `VectorizedArray<float, 16>`

compared to `std::exp` (both cases use a polynomial degree of five; details below).

The algorithm is based on the IEEE-754 floating-point representation

$$(-1)^s 2^{p-b} (1+m),$$

where $s$ is the sign bit, $p$ is the exponent, and $m$ is the mantissa with $0 \le m < 1$. $b$ is the bias, a constant depending only on the specific type.

For the exponential function one can write

$$exp⁡(x)=2^{xlog⁡2(e)}=2^y=2^{y_i} 2^{y_f}​$$

where $y = x \log_2(e)$, $y_i = \lfloor y \rfloor$ is the integer part, and $y_f = y - y_i$ is the fractional part. From the IEEE-754 expression one can identify

$$s=0,2^{y_i}=2^{p−b},2^{y_f}=1+m.$$

The integer exponent $y_i$ can be computed directly. However, although $y_f$ is known, reconstructing the mantissa

$$m=2^{y_f}-1$$

still requires evaluating $2^{y_f}$, which is transcendental and therefore expensive.

To avoid transcendental operations, a correction function

$$K(y_f) = 1 + y_f - 2^{y_f}$$

which is approximated using a polynomial of configurable degree, is introduced. With this approximation, the mantissa can be written as

$$m= y_f - K(y_f)$$

The accuracy of the exponential approximation depends primarily on the polynomial used for $K(y_f)$. The degree of the polynomial can be chosen using the template parameter `degree`. Supported degrees with precomputed coefficients are in the range [3, 11]. Higher degrees typically improve accuracy at the expense of additional compute cost. In many applications, a degree of 5 already yields an excellent balance between precision and performance, but users may tune it depending on their specific requirements.

Once the mantissa and exponent have been computed, the final floating-point value is assembled through internal bit manipulation that constructs the correct IEEE-754 representation. For further algorithmic details, see the referenced literature.

For evaluating the polynomial, I experimented with both Estrin's and Horner's schemes. To my surprise, benchmarks on my machine indicated that using Horner's scheme resulted in roughly 20% better overall performance compared to Estrin's scheme. Based on this result, I chose to go with Horner's scheme.

fyi @richardschu

kronbichler commented on include/deal.II/base/vectorization.h:
```diff
@@ -6983,4 +7087,452 @@ namespace std
+   * x^{\text{degree}}
+   * @f]
+   * where the coefficients @f$c_i@f$ are provided in ascending order using
+   * the Horner's scheme.
```
I would write
```suggestion
   * Horner's scheme.
```

kronbichler reviewed (approved):
Very nice, I like this suggestion a lot! I have a few comments (see below) and one comment: You say you tested the Horner scheme and Estrin's scheme and found it to be faster. Since you already provide many options and did test both cases, why not also add those 10 extra lines to do Estrin and let the user pick it via a template parameter? The reason I ask is that it depends on the code around the exp approximation whether the reduction of the latency chain by Estrin (at the expense of a log2(degree) more multiplications) is helpful. If you apply this function to an array of numbers that are all independent, I believe the out-of-order window in modern CPUs is long enough to overlap at least 4-6 of those, so you get pretty good performance. If you have a complicated algorithm around it doing many other things, it can be the latency-critical part (at least it was so often in my experiments). I did not test it, but I believe that starting with a number sufficiently close to 0 and computing `exp(exp(exp(exp(exp(x)))))` or something similar might expose different characteristics. In any case, I believe it is best if you choose.

richardschu commented on tests/base/vectorization_17.cc:
```diff
@@ -0,0 +1,101 @@
+#include <iomanip>
+#include <typeinfo>
+
+#include "../tests.h"
```
another comment regarding the test: for me looking at the potential use of that function it would be cool, if you would compare to the result obtained via `std::exp()` just to get an idea how accurate it is depending on `pol_degree` instead of outputting the resulting value.
Ofc the information is the same, but reading the `.output` one would directly see that the approximation is faithful when using a high enough polynomial degree for the correction factor.
And are the numbers in `test_param` kinda random or how did you select them?

richardschu commented on include/deal.II/base/vectorization.h:
```diff
@@ -6983,4 +7087,452 @@ namespace std
+   * This is log_2(e).
+   */
+  template <typename Number>
+  constexpr Number log2e = 1.44269504088896339;
```
oh I just saw there is already `LOG2E` in there ... with a different value --- can you check that?

gassmoeller reviewed (commented):
I like the addition and will most likely use it in the future as well, thanks for that (and I would also be very excited about `std::pow`!).

I have a question about the structure though. You introduced the function `fast_exp_approximation` in the general `dealii` namespace, and then introduced the namespace `dealii::internal::fast_transcendental` to store internal data. Since you are planning to add other functions as well, wouldnt it be cleaner to start a new namespace `dealii::fast_transcendental` for the functions (and maybe name the function `dealii::fast_transcendental::exp_approximation` or even `dealii::fast_transcendental::exp` since the difference to the normal `exp` is clear from the namespace name), and store the internal data in the namespace `dealii::fast_transcendental::internal`? That way we wouldnt increase the number of symbols in the general dealii namespace, and the namespace logic seems more similar to the rest of deal.II to me as well.

And I wonder if `vectorization.h` is the right place for your new functions. They happen to make use of vectorization, but the algorithm itself would also work without vectorization and would still be faster than a standard `std::exp`, correct? Then maybe it would be better to start a new header file, e.g. `fast_transcendental.h`, in particular since you plan to add other functions in the future this seems the more scalable option. This would also keep `vectorization.h` from growing further.

Both are just my thoughts, if anyone thinks the current structure works better let me know, this was just my impression.

bangerth commented on include/deal.II/base/vectorization.h:
```diff
@@ -25,6 +25,7 @@
 #include <algorithm>
 #include <array>
 #include <cmath>
+#include <cstring>
```
Do you actually need this include?

kronbichler replied:
I think it's needed for `memcpy` used below.

kronbichler commented on include/deal.II/base/vectorization.h:
```diff
@@ -6565,11 +6669,532 @@ namespace internal
+
+  VectorizedArray<Number, width> fractional_exponent = r - r.get_floor();
+
+  if (polynomial_evaluation_scheme ==
```
Can you use `if constexpr` here as well?

j-brotz commented:
Just as an idea, thinking about placing the implementation somewhere else, I am also wondering whether it would make sense to move the implementations of Horner’s and Estrin’s schemes out of the `internal` namespace and most likely also out of `fast_transcendetnal`. The schemes might be useful in many other situations independent of the transcendental function approximations.

gassmoeller commented:
> I measure a speedup between roughly 1.5 and 4 on my machine. In my opinion, that is sufficient justification to provide the implementation for pure double and float as well.

Thanks for checking! I did not necessarily mean that you have to include a non vectorized implementation as well (although it may be useful for some users so you are welcome to include it if it is not too much work). I was more thinking that conceptually what you propose is a fast approximation to transdendental functions, and just the fact that it happens to make use of vectorization does not mean it necessarily has to go into `vectorization.h` and having a new file, e.g. `fast_transcendental.h` would be easier to understand for the user.

> Just as an idea, thinking about placing the implementation somewhere else, I am also wondering whether it would make sense to move the implementations of Horner’s and Estrin’s schemes out of the internal namespace and most likely also out of fast_transcendetnal. The schemes might be useful in many other situations independent of the transcendental function approximations.

That may be the case and useful for the future, but I would say it depends on how concrete your plans are for using it in other places. I was mostly suggesting to start another namespace for the transcendental functions, because you already mentioned you have ideas/implementations for other functions as well, therefore it is already useful to create a proper namespace for them. Putting the Horner's and Estrin's scheme into an internal namespace as long as we do not have concrete plans for where else to use them is easier to change in the future, because (I think?) we do not guarantee compatibility for functions in `internal` namespaces. I.e. if we need them in other contexts we can always move them. Or were you thinking of an already existing namespace? Have you looked at the `Polynomial` namespace in `polynomial.h`? It looks to me like the class `Polynomial` already has an implementation of Horner's scheme, just for the more general case of polynomials with run-time selected degrees, so your implementations could go into `Polynomial` or `Polynomial::internal`.

j-brotz commented:
I extracted the `fast_transcendental` functionality into a new file `base/fast_transcendental.h` and updated the implementation so it now also supports `float` and `double`. From my side, the implementation should now be ready to be merged.

kronbichler commented:
Could you please squash you commits into a smaller number of self-contained commits, with the commits merely correcting something from a previous commit merged together? We strive to have a history of commits in deal.II that avoid the (local) development stages. If you have not done that before, see, e.g. https://stackoverflow.com/questions/14534397/squash-all-my-commits-into-one-for-github-pull-request (I assume you already know what we write on contributing, https://github.com/dealii/dealii/wiki/Contributing). Thank you!

bangerth commented on include/deal.II/base/vectorization.h:
```diff
@@ -6565,11 +6669,532 @@ namespace internal
+                    "Unsupported polynomial degree for exp_pol_coeff. "
+                    "Valid degrees are in the range [3, 11].");
+      return std::array<Number, degree + 1>{};
+    }();
```
Yes, I like this. This is a good way to disable the general template.

j-brotz replied:
Unfortunately, I had to revert the condition to `degree > 2 && degree < 12`. Setting it to `false` caused failures in the Jenkins serial and MPI check.

j-brotz commented:
Thanks, I hope the latest change fixes the test.

gassmoeller reviewed (approved):
Great, then this is good to merge. Thanks for the contribution, I am excited to give this a try.
